src/corecel/cont/Bitset.hh

```diff
@@ -0,0 +1,310 @@
+            }
+        }
+
+        return false;
+    }
+
+    CELER_CONSTEXPR_FUNCTION bool none() const noexcept { return !any(); }
+
+    CELER_CONSTEXPR_FUNCTION size_t count() const noexcept
+    {
+        size_t count = 0;
+        for (size_t i = 0; i < num_words; ++i)
+        {
+#if CELER_DEVICE_COMPILE
+            count += __popc(words_[i]);
+#elif defined(_MSC_VER)
+            count += __popcnt(words_[i]);
+#else
+            count += __builtin_popcount(words_[i]);
+#endif
+        }
+
+        return count;
+    }
+
+    CELER_CONSTEXPR_FUNCTION size_t size() const noexcept { return N; }
+
+    CELER_CONSTEXPR_FUNCTION Bitset& set() noexcept
+    {
+        for (size_t i = 0; i < num_words; ++i)
+        {
+            words_[i] = ~static_cast<word_type>(0);
+        }
+        // sanitize the last word
+        detail::Sanitize<N % bits_per_word>::sanitize(last_word());
+        return *this;
+    }
+
+    CELER_CONSTEXPR_FUNCTION Bitset& set(size_t pos, bool value = true) noexcept
+    {
+        if (value)
+        {
+            get_word(pos) |= mask(pos);
+        }
+        else
+        {
+            get_word(pos) &= ~mask(pos);
+        }
+
+        return *this;
+    }
+
+    CELER_CONSTEXPR_FUNCTION Bitset& reset() noexcept
+    {
+        for (size_t i = 0; i < num_words; ++i)
+        {
+            words_[i] = static_cast<word_type>(0);
+        }
+
+        return *this;
+    }
+
+    CELER_CONSTEXPR_FUNCTION Bitset& reset(size_t pos) noexcept
+    {
+        get_word(pos) &= ~mask(pos);
+        return *this;
+    }
+
+    CELER_CONSTEXPR_FUNCTION Bitset& flip() noexcept
+    {
+        for (size_t i = 0; i < num_words; ++i)
+        {
+            words_[i] = ~words_[i];
+        }
+        // sanitize the last word
+        detail::Sanitize<N % bits_per_word>::sanitize(last_word());
+        return *this;
+    }
+
+    CELER_CONSTEXPR_FUNCTION Bitset& flip(size_t pos) noexcept
+    {
+        get_word(pos) ^= mask(pos);
+        return *this;
+    }
+
+  private:
+    static CELER_CONSTEXPR_FUNCTION size_t which_word(size_t pos) noexcept
+    {
+        return pos / bits_per_word;
+    }
+
+    static CELER_CONSTEXPR_FUNCTION size_t which_bit(size_t pos) noexcept
+    {
+        return pos % bits_per_word;
+    }
+
+    static CELER_CONSTEXPR_FUNCTION word_type mask(size_t pos) noexcept
+    {
+        return static_cast<word_type>(1) << which_bit(pos);
+    }
+
+    CELER_CONSTEXPR_FUNCTION word_type get_word(size_t pos) const noexcept
+    {
+        return words_[which_word(pos)];
+    }
+
+    CELER_CONSTEXPR_FUNCTION word_type& get_word(size_t pos) noexcept
+    {
+        return words_[which_word(pos)];
+    }
+
+    CELER_CONSTEXPR_FUNCTION word_type& last_word() noexcept
+    {
+        return words_[num_words - 1];
+    }
+
+    CELER_CONSTEXPR_FUNCTION word_type last_word() const noexcept
+    {
+        return words_[num_words - 1];
+    }
+
+    word_type words_[num_words] = {};
```

Comment:
This class is no longer trivially constructable so I'm not sure we can have it as device data: should our `bitset` be an "operation class" that takes a Span or a pointer so that it can operate on preallocated data?